### ISSUE TITLE
Addition of meta fields for record information

### ISSUE THREAD

**Issue by siuc-nate** [CredentialEngine/CredentialRegistry#364]
We have encountered a use case where a consuming partner wants to be able to search specifically for records published by one or more given organizations. Currently we don't have an easy way to do this with the SPARQL search that I'm aware of, so what's the best way to accommodate that?  

I'm thinking it would be via another meta property or two, similar to some of the existing meta properties we already have for various other purposes.  This would be on all of the top level things (credentials, organizations, etc.) and use the CTID for the organization that actually owns the record as well as the CTID for the organization that actually published the record (i.e. not Credential Engine's CTID, but the organization itself) to accommodate third-party publishing cases.  Something like:

For records that are owned and published by the same organization:
```
{
  "credreg:__recordOwnedBy": "ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476",
  "credreg:__recordPublishedBy": "ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476"
}
```
```
<https://credentialengineregistry.org/resources/ce-235dc280-26b9-4ae0-97ef-591de0599d5f> credreg:__recordOwnedBy 'ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476' .
<https://credentialengineregistry.org/resources/ce-235dc280-26b9-4ae0-97ef-591de0599d5f> credreg:__recordPublishedBy 'ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476' .
```

For records that are owned by one organization and published by another (for third-party publishing use cases):
```
{
  "credreg:__recordOwnedBy": "ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476",
  "credreg:__recordPublishedBy": "ce-546a2c13-63c2-4bb1-88b5-31e0a74ce417"
}
```
```
<https://credentialengineregistry.org/resources/ce-235dc280-26b9-4ae0-97ef-591de0599d5f> credreg:__recordOwnedBy 'ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476' .
<https://credentialengineregistry.org/resources/ce-235dc280-26b9-4ae0-97ef-591de0599d5f> credreg:__recordPublishedBy 'ce-546a2c13-63c2-4bb1-88b5-31e0a74ce417' .
```

If there's an existing way to do this, or a better way to do this, I am open to suggestions.  The goal is to avoid having to get a record, then go get its envelope, then try to figure out which organization corresponds to the information in the envelope which might tell you the above.

Do those envelope records still indicate Credential Engine as being the "publisher" for everything just because the data always flows through our system?  That wouldn't be useful in this case. We want to be able to query on the actual owner/publisher of the data.

**Comment by stuartasutton** [comment]
@siuc-nate, can you please clarify the notion of "record"?

**Comment by siuc-nate** [comment]
In this case, "record" refers to the literal JSON-LD record itself in the registry.  The fields in this issue would be used for the SPARQL search, and wouldn't be part of the CTDL schema proper.

**Comment by stuartasutton** [comment]
Thanks for the clarification.

**Comment by excelsior** [comment]
I'm not sure as to how to obtain this metadata.

For envelopes published on behalf of an organization, we have the organization and the publisher. There is currently only one publisher in the system, I don't know its CTID, though.

How do I find out the CTID of the owner of an envelope published by its owner?

**Comment by siuc-nate** [comment]
I don't recall the reasoning behind it but I believe the system on the registry's end is setup so that CE is the only "publisher" - but if the data owner and data publisher (that is, the one or two other organizations besides CE involved) are not being provided as part of the publish operation, then we should provide that.

@cwd-mparsons @science Do you recall the details of this?

**Comment by siuc-nate** [comment]
Per our 10-1-2020 meeting
We will be responsible for providing this information, to be stored in the envelope and included in the index.

The properties to be used are:

- `documentOwnedBy` - The CTID of the intellectual property owner of the resource
- `documentPublishedBy` - The CTID of the organization that (via Credential Engine) published the resource to the Registry

**Comment by mparsons-ce** [comment]
@excelsior How will these new properties be provided in the publishing call?
Would additional parameters be added to the URL
https://credentialengineregistry.org/resources/organizations/{0}/documents?skip_validation=true&documentOwnedBy =ce-...&documentPublishedBy =ce-...

**Comment by excelsior** [comment]
@mparsons-ce @siuc-nate 

In the publish-on-behalf API:

```
POST https://credentialengineregistry.org/resources/organizations/{organization_id}/documents
```

`organization_id` is the internal ID (UUID) of an organization, not the CTID. Should we use internal IDs for the new properties as well for the sake of consistency? Or conversely, use the CTID in the above endpoint?

**Comment by excelsior** [comment]
@siuc-nate The RDF properties for owner and publisher metadata are `credreg:__recordOwnedBy` and `__recordPublishedBy `, respectively. Those are assigned to each resource subject, not the whole graph. The properties are omitted if an envelope lacks an owner or publisher, thus they should be optional in the queries.

**Comment by mparsons-ce** [comment]
I think it should be the CTID. 
We will want to be able to filter on these values and users will not know the internal ids.

**Comment by excelsior** [comment]
@mparsons-ce 

I changed the publishing APIs to use CTIDs instead of internal IDs. It's available on sandbox, the details are in [Swagger](https://sandbox.credentialengineregistry.org/swagger/index.html).

- `POST /envelopes?owned_by={owner_ctid}&published_by={publisher_ctid}`
- `POST /resources/{owner_ctid}/documents?published_by={publisher_ctid}`

Note that the owner CTID is located in the path of the latter endpoint and is required. In the other cases, it's optional.

**Comment by mparsons-ce** [comment]
@excelsior 
I want to confirm whether the publish endpoint has changed. 
In the above you show the POST as:
`POST /resources/{owner_ctid}/documents?published_by={publisher_ctid}`
The swagger docs show the **/organization/**, still in the path. Should **/organization/** still be in the endpoint, or is it do be removed?
As well Swagger still show {organization_id} as the parameter name (which of doesn't really matter as it will be a string).

![image](https://user-images.githubusercontent.com/58825866/96773966-edcdfd00-13aa-11eb-853a-a8a68e512e39.png)

**Comment by excelsior** [comment]
@mparsons-ce Yes, the `organizations` part is still there, the info in Swagger is correct. The meaning behind the `organization_id` parameter is what has been changed. Sorry about the confusion.

**Comment by mparsons-ce** [comment]
@excelsior 
I tried doing a sandbox publish using the new pattern of the org ctids:
https://sandbox.credentialengineregistry.org/resources/organizations/ce-b67e212e-2eb7-401e-a076-bf566102f268/documents?skip_validation=true&published_by=ce-a4041983-b1ae-4ad4-a43d-284a5b4b2d73

I get the following error:
**{"errors":["Couldn't find Organization"]}**

**Comment by mparsons-ce** [comment]
@excelsior The latter organization may be one with issues. I will try others. 
Same error with:
https://sandbox.credentialengineregistry.org/resources/organizations/ce-5ea303a9-0cba-42db-b828-38d2f1fb890d/documents?skip_validation=true&published_by=ce-a4041983-b1ae-4ad4-a43d-284a5b4b2d73

![image](https://user-images.githubusercontent.com/58825866/96937499-e6cbeb00-148d-11eb-8415-13aaa362347e.png)

**Comment by excelsior** [comment]
@mparsons-ce The problem is caused by the publishing organization specified in those requests: there is no organization with CTID `ce-a4041983-b1ae-4ad4-a43d-284a5b4b2d73` on sandbox.

**Comment by jeannekitchens** [comment]
@mparsons-ce what is the status of this.  Please update ahead of this Thursday.

**Comment by siuc-nate** [comment]
@mparsons-ce will test this.

**Comment by jeannekitchens** [comment]
@mparsons-ce please go through these issues that were identified to be tested last week so we can take care of open issues from 2020 and close them.

**Comment by mparsons-ce** [comment]
@excelsior 

I published a record and don't see the envelope metadata additions.
Publishing URL
https://sandbox.credentialengineregistry.org/resources/organizations/ce-72d32373-98db-4201-b9ae-59e78df75c4e/documents?skip_validation=true&published_by=ce-72d32373-98db-4201-b9ae-59e78df75c4e
Envelope:
https://sandbox.credentialengineregistry.org/envelopes/ce-b9e84127-f4a0-4405-9e0d-516197ccecd4

![image](https://user-images.githubusercontent.com/58825866/99705758-c88ee600-2a5f-11eb-87b6-bc5f5c352a0a.png)

**Comment by excelsior** [comment]
@mparsons-ce The publishing metadata was not part of the node headers, I fixed that.

**Comment by mparsons-ce** [comment]
Once the metadata is added to the envelope directly, how can we search for a publisher? 
For example, would this work:
https://sandbox.credentialengineregistry.org/ce-registry/search?published_by=ce-89bd0d16-6492-4ae1-8059-deabc3d89c15

As a test I tried searching for other envelope properties and was not successful. For example the following didn't return any data:
https://sandbox.credentialengineregistry.org/ce-registry/search?envelope_ceterms_ctid=ce-10a689e8-a40b-46e8-8059-c6cb66d04eae
OR
https://sandbox.credentialengineregistry.org/ce-registry/search?envelope_id=002c72ec-5832-4816-b539-2ff8206eecb0

**Comment by excelsior** [comment]
@mparsons-ce The envelope metadata is exposed as top-level attributes now on sandbox.

The search mechanism only allows to perform full-text queries against the indexed attributes (those defined in the config). If you need a faceted search, just let me know which attributes you'd like to apply filters to, and I'll expand this functionality to support it. Let's create a separate issue for that task.

**Comment by mparsons-ce** [comment]
@excelsior 
I still don't see the metadata added in the sandbox:
![image](https://user-images.githubusercontent.com/58825866/102421945-45fc3680-3fcb-11eb-92c5-7e657b6e5fba.png)

![image](https://user-images.githubusercontent.com/58825866/102422052-852a8780-3fcb-11eb-8e20-6828f8adc033.png)

**Comment by excelsior** [comment]
@mparsons-ce That's weird. They're there, closer to the end, right before the `changed` property:

https://sandbox.credentialengineregistry.org/envelopes/ce-f3a69a85-b168-404f-96b1-85d3da7e1718
https://sandbox.credentialengineregistry.org/envelopes/ce-4602d328-4a38-42d0-b6b8-3a8de4b8f4ed

Which endpoint are you using?

**Comment by mparsons-ce** [comment]
@excelsior I was using the sandbox search. 
https://sandbox.credentialengineregistry.org/ce-registry/search?resource_type=transfer_value_profile

**Comment by mparsons-ce** [comment]
How would I search for published_by? The following doesn't return results:
https://sandbox.credentialengineregistry.org/ce-registry/search?published_by=ce-a4041983-b1ae-4ad4-a43d-284a5b4b2d73

**Comment by excelsior** [comment]
@mparsons-ce The metadata fields were added to search results as well on sandbox.

**Comment by mparsons-ce** [comment]
Thanks @excelsior This works on the sandbox.

I see that on production, the metadata properties appear on the envelope. However the search has not been enabled (which would make sense since we have just verified this in the sandbox). 
If it has been enabled in production, it doesn't work: 
https://credentialengineregistry.org/ce-registry/search?published_by=ce-e4c8e9bd-88bd-4704-9e3a-78faebdca59d

In summary, please now enable this feature on staging and production.
Thanks

**Comment by mparsons-ce** [comment]
@excelsior This issue started with a request by Nate to be able to filter on the owned_by/published_by in SparQL searches. 
Does this update include populating the SparQL indices appropriately?
Nate has used the following as an example:
```
{
  "credreg:__recordOwnedBy": "ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476",
  "credreg:__recordPublishedBy": "ce-c5e501ce-11fd-4d72-a5ad-4b49843f5476"
}
```
A later post changed the properties
```
The properties to be used are:
documentOwnedBy - The CTID of the intellectual property owner of the resource
documentPublishedBy - The CTID of the organization that (via Credential Engine) published the resource to the Registry
```

We then finalized the envelope properties to be:
```
- owned_by
- published_by
```

SO:
Have the owned_by/published_by been added to SparQL?
if so, what are the property names?

**Comment by excelsior** [comment]
@mparsons-ce @siuc-nate The metadata has been available in SPARQL for a long time already. The properties have the originally suggested names: `credreg:__recordOwnedBy` and `credreg:__recordPublishedBy`.

**Comment by mparsons-ce** [comment]
Thanks, closing this issue.
